src/aspire/volume/__init__.py

```diff
@@ -1,16 +1,164 @@
+from aspire.utils.preprocess import downsample
 
 
 class Volume:
     """
-    stack of volumes
+    Volume is an N x L x L x L array, along with associated utility methods.
     """
+    def __init__(self, data):
+        """
+        Create a volume initialized with data.
+
+        Volumes should be N x L x L x L,
+        or L x L x L which implies N=1.
+
+        :param data: Volume data
+
+        :return: A volume instance.
+        """
+
+        if data.ndim == 3:
+            data = data[np.newaxis, :, :, :]
+
+        ensure(data.ndim == 4,
+               'Volume data should be ndarray with shape NxLxLxL'
+               ' or LxLxL.')
+
+        ensure(data.shape[1] == data.shape[2] == data.shape[3],
+               'Only cubed ndarrays are supported.')
+
+        self.data = data
+        self.N = self.data.shape[0]
+        self.dtype = self.data.dtype
+        self.resolution = self.data.shape[1]
+        self.shape = self.data.shape
+        self.volume_shape = self.data.shape[1:]
+
+    def __getitem__(self, item):
+        # this is one reason why you might want Volume and VolumeStack classes...
+        #return Volume(self.data[item])
+        return self.data[item]
+
+    def __setitem__(self, key, value):
+        self.data[key] = value
+
+    def __len__(self):
+        return self.N
+
+    def __add__(self, other):
+        return Volume(self.data + other.data)
+
+    def __sub__(self, other):
+        return Volume(self.data - other.data)
+
+    def __mul__(self, other):
+        if isinstance(other, Volume):
+            res = Volume(self.data * other.data)
+        else:
+            res = Volume(self.data * other)
+
+        return res
+
+    # def __truediv__(self, other):
+    #     if isinstance(other, Volume):
+    #         res = Volume(self.data / other.data)
+    #     else:
+    #         res = Volume(self.data / other)
+
+    #     return res
+
+    # # TODO: Open issue for fixing rotated_grids wholesale
+    def _rotated_grids(self, rot_matrices):
+        raise NotImplementedError('Method not currently implemented')
+
+    def project(self, vol_idx, rot_matrices):
+        data = self[vol_idx].T  #RCOPT
+
+        n = rot_matrices.shape[0]
+
+        pts_rot = rotated_grids(self.resolution, rot_matrices)
+
+        # TODO: RCOPT come back and convert these methods internally to C order
+        ## TODO: rotated_grids might as well give us correctly shaped array in the first place,
+        pts_rot = m_reshape(pts_rot, (3, self.resolution**2*n))
+
+        im_f = (1./self.resolution *
+                Plan(self.volume_shape, pts_rot).transform(data))
+
+        im_f = im_f.reshape(-1, self.resolution, self.resolution)
+
+        if self.resolution % 2 == 0:
+            im_f[:, 0, :] = 0
+            im_f[:, :, 0] = 0
+
+        im_f = centered_ifft2(im_f)
+
+        # Looks like we still need to transpose the image axis here for repro
+        im_f = np.swapaxes(im_f, -2, -1)         # RCOPT
```

Comment:
This is a bit of a shot in the dark, but the `swapaxes` here may be due to the use of `m_reshape` above (with respect to the spatial positions, not the image stack index).

Reply:
Correct, I was able to change the axis of the rotated_grids `(k,l1,l2,n)~~>(k,l2,l1,n)`.  For now this will at least push the change away from the fft.  I think `rotated_grids` is slated to be improved anyway...

Reply:
Good. Agreed that `rotated_grids` needs to be fixed.